screeps-game-api/src/objects/impls/structure_controller.rs

```diff
@@ -0,0 +1,58 @@
+use stdweb::Value;
+
+use objects::StructureController;
+
+simple_accessors! {
+    StructureController;
+    (level -> level -> u32),
+    (progress -> progress -> u32),
+    (progress_total -> progressTotal -> u32),
+    (safe_mode -> sameMode -> u32),
+    (safe_mode_available -> sameModeAvailable -> u32),
+    (safe_mode_cooldown -> sameModeCooldown -> u32),
+    (ticks_to_downgrade -> ticksToDowngrade -> u32),
+    (upgrade_blocked -> upgradeBlocked -> u32)
+}
+
+#[derive(Debug)]
+pub struct Reservation {
+    pub username: String,
+    pub ticks_to_end: u32
+}
+
+#[derive(Debug)]
+pub struct Sign {
+    pub username: String,
+    pub text: String,
+    pub time: u32,
+    pub datetime: String  // todo: use real date type
+}
+
+impl StructureController {
+    pub fn reservation(&self) -> Option<Reservation> {
+        if let Value::Reference(r) = js!(return @{self.as_ref()}.reservation;){
```

Comment:
For style I'd use a `match` with a catch-all `_` branch rather than `if let .. else`, but that's not a big deal.

Reply:
Any reasons why? Still learning my way around Rust and that line was harder to write and debug than I'd care to admit.

The thing I could see is being able to extend match going forward.